### PR TITLE
Refactor rhythm playback pipeline and loader

### LINKED ISSUE #647: Full ABC tunes (pattern_type = "Full track") don't display or play correctly in the Rhythm Player
## Description

Full ABC tunes with `pattern_type = "Full track"` don't display or play correctly in the Rhythm Player.

### Symptoms
1. **Rendering issue**: The ABC notation renders on a single line instead of respecting the multi-line layout of the tune.
2. **Repeat markers ignored**: The playback does not respect repeat signs (`:|`, `|:`) in the ABC notation.

### Likely Cause
The Rhythm Player's ABC rendering or parsing logic may not be handling full multi-part tunes correctly — it may be treating them like short rhythmic patterns rather than full musical compositions with standard repeat structures.

### Context
The `pattern_type` column in `rhythm_patterns` supports multiple types, with `"Full track"` being intended for complete tunes with repeats and multi-line structure. Other types (e.g., short loop patterns) work correctly.

### Related
- Part of the broader rhythm accompaniment feature (#607)
- Related to pattern-type handling in `RhythmService.ts` and ABC rendering in `RhythmPlayer.tsx`

### PR DESCRIPTION
**Pull request by sboagy**
## Summary
- extract the rhythm playback pipeline into dedicated loader, canonical ABC, notation ABC, playback planning, structured playback model, and structured display-sync modules
- add focused tests around the refactored pipeline and keep the player/service behavior aligned with the new module boundaries
- fix rhythm note highlighting regressions for structured section playback and rest-heavy wrapped notation lines
- follow through on the architecture review by removing duplicated helpers/types, moving notehead-resolution logic into the display-sync module, and updating the permanent development doc

Fixes https://github.com/sboagy/tunetrees/issues/647 .

## What Changed
- extracted `loadRhythmPattern(...)` behavior out of `RhythmService.ts` into `pattern-loader.ts`
- extracted canonical ABC assembly into `canonical-abc.ts`, display-safe notation assembly into `notation-abc.ts`, and playback planning into `playback-abc.ts`
- extracted pure runtime section/beat logic into `structured-playback-model.ts`
- extracted notehead and label synchronization logic into `structured-display-sync.ts`
- added shared rhythm pattern types in `src/lib/rhythm/pattern-types.ts`
- added shared tune-type alias/lookup normalization in `src/lib/rhythm/tune-type-lookup.ts`
- removed the local `normalizeStructure` shadow copy from `RhythmPlayer.tsx`
- moved `resolveCurrentBeatNotehead` and `resolvePlaybackMarkerNotehead` out of `RhythmPlayer.tsx` and into `structured-display-sync.ts`
- corrected the permanent architecture/dev doc in `docs/development/rhythm-player-and-pattern-editor.md` so it matches the current implementation and review feedback

## Bug Fixes Included
- fixed structured rhythm section note highlighting so section playback resolves the correct noteheads
- fixed the rest-heavy line-wrap bug where full-track playback could highlight the wrong noteheads after notation wrapped across lines
- added/updated regression coverage for the structured playback and wrapped-line synchronization paths

## Review Response
- removed stale comments that described pre-extraction architecture
- deduplicated repeated tune-type lookup tables across loader, service, and pattern query code
- unified `RhythmPatternRequest` and `RhythmPatternMetadata` so service and loader share the same source of truth
- added inline guidance in the notehead fallback logic where the structured/direct/template/line fallback tiers are hardest to read
- documented the `loadedPattern() ?? metadata()` duality and the `resolveEffectivePlaybackParts(...)` helper in the permanent doc

## Validation
- `npm run lint`
- `npm run check`
- `npm run typecheck`
- `npm run codegen:schema:check`
- `npm run test:unit`
- focused tests: `src/components/rhythm/RhythmPlayer.test.tsx` and `src/lib/services/RhythmService.test.ts`

## Commits
- `2fcf23af` refactor: extract rhythm abc pipeline
- `b784202c` refactor: split rhythm pattern loading
- `2b105ada` test: add structured playback browser regression
- `11fba604` Fix rhythm section note highlighting
- `d4aff1e4` Fix rhythm note highlighting across rest-heavy line wraps
- `4dbabb7f` Refine rhythm review followups and docs

### CODE REVIEW

File: src/lib/rhythm/canonical-abc.ts [hunk 1]
```diff
@@ -0,0 +1,568 @@
+import type { RhythmPatternType } from "@/lib/services/RhythmService";
```
[the remark this answers is not in the document]
Comment (sboagy) [reply src/lib/rhythm/canonical-abc.ts]: Addressed in 4dbabb7f. `RhythmPatternType` now lives in `src/lib/rhythm/pattern-types.ts`, and `src/lib/rhythm/canonical-abc.ts` imports it from the rhythm-domain module instead of from `RhythmService.ts`.

File: src/lib/rhythm/pattern-loader.ts [hunk 2]
```diff
@@ -0,0 +1,831 @@
...
+  genreId?: string | null;
+  genreName?: string | null;
+  tuneTypeName?: string | null;
+  tuneId?: string | null;
+  userId?: string | null;
+  selectedPatternId?: string | null;
+}
+
+export type RhythmPatternType = "seed" | "full_track";
+
```
[the remark this answers is not in the document]
Comment (sboagy) [reply src/lib/rhythm/pattern-loader.ts]: Addressed in 4dbabb7f. The shared rhythm pattern request/metadata/candidate types now live in `src/lib/rhythm/pattern-types.ts`; `pattern-loader.ts` re-exports them and `RhythmService.ts` re-uses the same definitions instead of keeping parallel copies.

File: e2e/tests/rhythm-player-001-structured-playback.spec.ts [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+      // visible viewport. The regression should fail immediately the first time
+      // the blue note jumps backward on the same rendered line before playback
+      // has traversed the later notation lines.
+      for (let attempt = 0; attempt < 160; attempt += 1) {
+        const snapshot = await readActiveNotationSnapshot();
+        if (snapshot) {
+          const wrappedOnSameLine =
+            lastSnapshot &&
+            snapshot.lineIndex === lastSnapshot.lineIndex &&
+            snapshot.x < lastSnapshot.x - 8;
```
[the remark this answers is not in the document]
Comment (sboagy) [reply e2e/tests/rhythm-player-001-structured-playback.spec.ts]: Addressed in 588f6fd1. The structured-playback browser regression no longer uses a fixed 160x150ms observation cap; it now derives the wait budget from the selected tempo, rendered note count, and repeated-structure factor, with bounded CI padding.

**Comment by sboagy** [comment]
Adding a visible summary here because the earlier responses were posted as inline replies on the review threads under Files changed.

Copilot review responses addressed:

- canonical-abc service-layer type dependency: addressed in `4dbabb7f` by moving `RhythmPatternType` into `src/lib/rhythm/pattern-types.ts` and importing it from the rhythm domain module instead of `RhythmService.ts`. Inline thread: https://github.com/sboagy/tunetrees/pull/660#discussion_r3290769772
- duplicated rhythm pattern types between loader and service: addressed in `4dbabb7f` by introducing shared rhythm pattern types in `src/lib/rhythm/pattern-types.ts`, re-exporting them from `pattern-loader.ts`, and reusing them from `RhythmService.ts`. Inline thread: https://github.com/sboagy/tunetrees/pull/660#discussion_r3290770511
- structured playback E2E observation budget too short: addressed in `588f6fd1` and formatted in `41f99fad` by replacing the fixed polling cap with a derived wait budget based on tempo, rendered note count, and repeated-structure factor. Inline thread: https://github.com/sboagy/tunetrees/pull/660#discussion_r3290770817

Validation after the E2E follow-up: `npm run check` and `npm run lint` passed.
